[yell.ow] 지하철 노선도 step3 코드리뷰 요청드립니다.

isangu opened the pull request:
이번 지하철 노선도 미션을 진행한 옐입니다. 많은 피드백 부탁드립니다 !! 😅

## 질문사항
1. 다양한 Exception 처리를 위해 ExceptionAdvise를 구현하는 방식과 Controller 마다 구현하는 방식 중 어느 방식이 주로 사용될까요? 선호되는 방식이 있는지 궁금합니다.
2. DAO 는 서비스를 위해서만 사용해야 하는지 궁금합니다. 오늘 수업을 들으며 도메인 객체에 비지니스 로직을 구현하려고 했지만, Line, Section, Station과 같은 객체들이 DAO를 직접 가지지 않고서는 이를 구현할 수 있는 방법이 마땅히 떠오르지 않았습니다. 도메인에서도 DAO를 직접 가지고 호출하는 방식으로 구현하는 것이 맞는걸까요?
3. jdbcTemplate으로 리팩터링을 진행하며 StationController —> StationService —> StationDao의 계층 구조를 가지려고 했는데, StationService가 단순히 StationDao 함수를 리턴해주는 역할밖에 수행하지 않게 구현이 되었습니다. 이런 경우에도 Controller가 Service를 호출하도록 설계하는게 적절할까요?
4. Request 객체나 response의 객체를 생성하거나, 사용하는 위치도 고려되어야 할까요? controller에서 request를 모두 읽고, controller에서 response를 생성해야 한다고 생각이 들었는데, service나 repository까지 전달해 사용이 가능할지가 궁금합니다.

kakao-yell-ow committed:
docs : README.md 요구사항 정리 추가

kakao-yell-ow committed:
feat : StationController의 ShowStations 및 deleteStation 구현

StationDao의 findAll을 이용하여 ShowStations 기능 구현
deleteStation에서 제거하려는 id가 존재하지 않는 경우 badRequest를 리턴하도록 수정

kakao-yell-ow committed:
feat : Line 및 LineController, LineDao 구현

LineController의 createLine 기능 구현

kakao-yell-ow committed:
feat : LineController, LineDao 리팩터링 및 DuplicateLineNameException 생성

중복된 이름의 Line을 추가하려고 할 경우 Custom Exception을 throw 하도록 수정
Linecontroller의 createLine 및 LineDao의 save 리팩터링

kakao-yell-ow committed:
feat : LineController의 노선 조회 기능 추가

특정 id의 노선을 찾는 showLine 및 모든 노선을 조회하는 showLines 기능 추가

kakao-yell-ow committed:
refactor : StationDao, StationController 리팩터링

중복된 Station 입력 시 BadRequest를 Response 하도록 리팩터링

kakao-yell-ow committed:
feat, docs : 노선 수정 기능 추가 및 README.md 오타 수정

LineController 및 LineDao의 노선 수정을 위한 기능 추가
README.md의 지하철 조회/수정/삭제 로 되어있던 부분을 노선 조회/수정/삭제로 수정
LineAceeptanaceTest의 BeforeEach에 static으로 선언된 LineDao 및 StationDao 초기화 코드 삽입

kakao-yell-ow committed:
feat : 노선 삭제 기능 추가

waylon-n committed:
refactor: Exception 발생 리팩터링

Exception 발생 시 Controller의 ExceptionHandler가 처리

waylon-n committed:
docs: README.md 2단계 기능목록 추가

waylon-n committed:
feat: 노선 내 정보 추가 및 구간 등록 기능

노선 추가 시 상행종점, 하행종점, 두 종점 간 거리 입력 요구
노선 Response에 노선 내 모든 지하철역 포함
노선 구간 등록 기능 추가 (4가지 상황에 따라 enum 사용)

waylon-n committed:
refactor: 노선 내 지하철역 목록 반환 리팩터링

waylon-n committed:
refactor: Line 리팩터링

waylon-n committed:
feat: 노선 내 지하철역 삭제 기능 추가

Section 등록 기능 Fix 필요

kakao-yell-ow committed:
fix : Line의 지하철 구간 추가 로직 수정

Line의 지하철 구간 삽입을 AddStatus를 사용하지 않고 작동하게끔 수정

kakao-yell-ow committed:
refactor : Line의 불필요 sout 코드 삭제

kakao-yell-ow committed:
fix : Line의 구간 추가에서 distance를 검증하지 않는 부분 수정

kakao-yell-ow committed:
refactor : Line의 메서드 분리를 통한 리팩터링 진행

waylon-n committed:
feat: JDBC 초기 설정

Dependency 추가
H2 DB 설정
DB 데이블 스키마 추가

waylon-n committed:
refactor: Station에 JDBC 적용

Dao 객체에 저장하던 데이터를 DB에 저장
StationService 추가

waylon-n committed:
refactor: Line, Section에 JDBC 적용

Controller, Service, Repository로 분리
Line 정보에 시작, 종착역 추가

kakao-yell-ow committed:
refactor : 사용하지 않는 AddStatus 삭제 및 LineService의 의존성 주입을 생성자 주입으로 변경

kakao-yell-ow committed:
refactor : 전체 로직 리팩터링

LineService 메서드 리팩터링 및 로직 분리
Line에 메서드 추가
함수 위치 변경

kakao-yell-ow committed:
refactor : Line의 생성자 this를 호출하도록 수정

kakao-yell-ow committed:
refactor : Section의 validateDistance 부등호 수정

kakao-yell-ow committed:
refactor : LineController 및 StationController RequestMapping Annotation 추가

Hue9010 commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,70 @@
+            lineId = jdbcTemplate.queryForObject("select id from LINE where name = ?", Long.class, lineRequest.getName());
+        } catch (Exception e) {
+            throw new DuplicateLineNameException("중복된 이름의 노선입니다.");
+        }
```
예외 처리 👍 
Exception을 catch해서 중복된 이름이 아닐 때도 DuplicateLineNameException이 발생한다는 문제가 있긴하지만 그에 따른 문제가 있다는 것만 인지하는 정도로 충분 해 보입니다. 😉

Hue9010 commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public boolean deleteById(Long id) {
+        if(stationDao.deleteById(id) == 1) {
```
1이 무엇을 의미할까요? 🤔 
삭제 실패라면 dao 내에서 예외 발생으로 처리 할 수도 있을 것 같아요.

Hue9010 commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,32 @@
+
+    public Station findById(Long id) {
+        return stationDao.findById(id);
+    }
```
사용되는 곳이 없네요. 불필요해진 코드는 불필요해진 시점에 제거해주세요. 😉

Hue9010 commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,86 @@
+    @ExceptionHandler({InvalidLineArgumentException.class, DuplicateLineNameException.class})
+    public ResponseEntity<String> badRequestErrorHandler(RuntimeException e) {
+        return ResponseEntity.badRequest().body(e.getMessage());
+    }
```
> 1.다양한 Exception 처리를 위해 ExceptionAdvise를 구현하는 방식과 Controller 마다 구현하는 방식 중 어느 방식이 주로 사용될까요? 선호되는 방식이 있는지 궁금합니다.

각 Controller마다 처리하게 되면 중복된 예외 처리가 발생하여 보통 ControllerAdvice를 만들어 처리하는 편입니다. 
[Spring Handle Exception](https://jaehun2841.github.io/2018/08/30/2018-08-25-spring-mvc-handle-exception/#%EB%93%A4%EC%96%B4%EA%B0%80%EB%A9%B0)에 정리가 잘 되어 있네요. 3번 리졸버는 리졸버라는 단어가 익숙치 않다면 당장은 신경쓰지 마세요. 😉 

중복된 예외처리의 예를 들자면 우리가 잘 알 고있는 IllegaArgumentException, RuntimeException 등에 대한 처리도 각 Controller마다 해줘야 합니다. 그리고 프로젝트가 커져서 더 다양한 도메인 생기게 된다면 InvalidXxxArgumentException이 만들어 질 때마다 핸들링을 해줘야할텐데 처음부터 
```java
public class InvalidLineArgumentException extends InvalidArgumentException {

    public InvalidLineArgumentException(String message) {
        super(message);
    }
}
```
InvalidArgumentException를 확장하여 사용한 다음 ControllerAdvice에서 InvalidArgumentException를 통해 공통적으로 핸들링 할 수 있답니다.

지금은 에러를 다루시는 것만해도 대단하신거라 지금 처리 해주신 것만으로도 충분합니다. 😉
다른 것들부터 먼저하고 여유 되신다면 부수적으로 ControllerAdvice을 사용해보는 것도 좋을 것 같네요.

Hue9010 commented on src/main/java/subway/line/Line.java:
```diff
@@ -0,0 +1,68 @@
+
+import java.util.Objects;
+
+public class Line {
```
> 2.DAO 는 서비스를 위해서만 사용해야 하는지 궁금합니다. 오늘 수업을 들으며 도메인 객체에 비지니스 로직을 구현하려고 했지만, Line, Section, Station과 같은 객체들이 DAO를 직접 가지지 않고서는 이를 구현할 수 있는 방법이 마땅히 떠오르지 않았습니다. 도메인에서도 DAO를 직접 가지고 호출하는 방식으로 구현하는 것이 맞는걸까요?

도메인에서  DAO를 직접 사용하는 부분이 없는데 해결된 걸까요? 🤔
아직 해결되지 않았다면 코드상 남아있는 부분이 없어 어떻게 어려움을 느꼈는지 확실치 않네요. 어떤부분이 어려웠는지 알려주시면 답변 드리도록 할게요. 🙏

Hue9010 commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,32 @@
+    @Autowired
+    private StationDao stationDao;
+
+    public Station save(Station station) {
+        return stationDao.save(station);
+    }
```
> 3. jdbcTemplate으로 리팩터링을 진행하며 StationController —> StationService —> StationDao의 계층 구조를 가지려고 했는데, StationService가 단순히 StationDao 함수를 리턴해주는 역할밖에 수행하지 않게 구현이 되었습니다. 이런 경우에도 Controller가 Service를 호출하도록 설계하는게 적절할까요?

대표적으로 여기가 그러겠네요. 답부터 드리자면 넵, 말씀처럼 Service를 이용하는게 좋습니다. 가장 단순한 이유를 먼저 보면 `Controller > Service > Dao 계층 구조로 설계`하셨기 때문입니다.

이런 구조를 지키는건 스프링의 장점이자 단점이라고 생각됩니다. 모두가 익숙한 구조를 사용하기 때문에 새로운 프로젝트를 분석 할 일이 있을 때 공통 된 구조로 인해 처음보는 프로젝트라도 빠르게 파악할 수 있습니다. 단점이라면 불필요하다고 느끼는 코드들이 생길 수 있습니다.(이건 스프링 + 자바의 단점이라고 볼 수도 있네요.)

각 layer별로 두고 보면 layer별로 지켜진 역할을 수행함으로서 Controller, Service, Dao에 어떤 내용이 있을지 예측 가능하기 때문에 앞서 말한 코드 파악에 유리합니다. 그 외에도 각 layer 별로 조금씩 예외를 두기 시작하면 조금씩 모든 걸 처리하는 만능 클래스가 생겨버리기 시작합니다. 이전 로또나 레이싱의 MVC를 생각해보자면 View에서 모든 걸 다 처리하는 상황으로 생각해 볼 수 도 있겠네요.

Hue9010 commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,70 @@
+        return line;
+    };
+
+    public Long save(LineRequest lineRequest) {
```
> 4. Request 객체나 response의 객체를 생성하거나, 사용하는 위치도 고려되어야 할까요? controller에서 request를 모두 읽고, controller에서 response를 생성해야 한다고 생각이 들었는데, service나 repository까지 전달해 사용이 가능할지가 궁금합니다.


![layer](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FbFruEV%2FbtqAUv4HJLQ%2FH5TVBjqkKc5KBgD4Vdyvkk%2Fimg.png)
출처 : [[스프링] Spring 웹 계층](https://leveloper.tistory.com/14) 

[테스트로 배우는 스프링 Core](https://edu.nextstep.camp/s/bb4PXLji/ls/dRX6Ioea) 강의 자료 하단에 있는 그림으로 찾았습니다. 😉 

Request, Response의 이름의 DTO입니다. Data Transfer Object로 데이터 전송을 위한 객체이지 저장을 위한 객체가 아닙니다. 그래서 지금처럼 DAO를 통해 저장하는 행위는 DTO의 사용 범위를 벗어난 행위입니다.

DTO를 domain으로 언제 변환하는 지에 대해서는 Controller -> Service도 서로 다른 layer이기 때문에 계층 간 데이터 전달을 위해 엄격하게 DTO를 사용해야한는 분들은 Service에서까진 Dto로 전달 받고 변환해서 사용하고, 그게 아니라면 Controller에서 Request -> domain -> response 변환을 합니다.

Hue9010 commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -0,0 +1,130 @@
+    public void updateDownStation(Long sectionId, Long newStationId, int newDistance) {
+        String sql = "update section set down_station_id = ?, distance = ? where id = ?";
+        jdbcTemplate.update(sql, newStationId, newDistance, sectionId);
+    }
```
사용되지 않고 있는 걸로 보이네요. 사용하지 않는 코드는 바로 지우는게 중요한게 이번 과제를 진행하면서도 여러 곳에서 사용하지 않는 코드가 남게 되었는데 수년동안 서비스 되고 있는 코드에 사용되지 않는 코드들이 그대로 남아 있다면... 어마 무시하겠죠? 😱
나중에 필요하지 않을까 싶은 코드도 가능하다면 git을 믿고 지우시는 걸 추천드립니다.

Hue9010 commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -0,0 +1,130 @@
+    public void updateUpStation(Long sectionId, Long newStationId, int newDistance) {
+        String sql = "update section set up_station_id = ?, distance = ? where id = ?";
+        jdbcTemplate.update(sql, newStationId, newDistance, sectionId);
+    }
```
사실 지금과 같이 단순한 쿼리도 거리와 상행을 저장한다는 로직이 쿼리에 존재하는 거랍니다. 😋
특히나 객체보다는 데이터 중심적인 로직입니다. DB에도 객체를 쓰고 읽는 형태로 처리해주세요. 여럽게 생각 하실 필요 없이 update, insert는 객체를 저장하고 갱신하게, select은 간단한 조건들로만 객체를 조회하게 사용해 주시면 됩니다.

예를 들면 지금은 사용되지 않고 있지만 바로 위에 있는 findDistanceById처럼 객체가 아닌 데이터를 조회하는 부분들은 제거해 주세요.

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+
+    private Section updateSection(Long lineId, SectionRequest sectionRequest) {
+        Long sectionId = sectionDao.findSectionIdFromEqualUpStationId(lineId, sectionRequest.getUpStationId());
+        if (sectionId == 0L) {
```
0L이 무엇을 의미하는지 상수로 표현해주면 좋을 것 같아요.

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+        Section section = sectionDao.findById(sectionId);
+        return new Section(sectionId, lineId, sectionRequest.getDownStationId(), section.getDownStationId(),
+                section.getDistance() - sectionRequest.getDistance());
+    }
```
DB를 통해 데이터를 조회하고 처리하고 있는데 조금은 비효율적으로 느낄 수 있지만 Line에 속하는 List<Section>을 조회해서 객체가 로직을 담당하게 해보세요.😉  필요하다면 Sections라는 일급 컬렉션을 만들 수도 있을 것 같아요.

지금 코드의 단점이라면 DAO부터 DB의 데이터까지 모두 인지를 하고 어떤 데이터를 어떻게 조작하고 있는지 분석을 해야합니다.

Hue9010 commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -0,0 +1,130 @@
+                "select * from findAllSections",
+                stationRowMapper, lineId, startStationId
+        );
+    }
```
List\<Section>을 조회하지만 id, lineId, distance가 비어있습니다. findAllSections을 통해 Section이 조회됐다고 기대하고 다른 곳에서 잘못 사용하여 NPE나 버그가 발생할 여지가 큽니다.
쿼리를 통해 처리하기보다는 findByLineId로 전체를 조회한 다음 자바 코드 상에서 정렬을 해주세요.

Hue9010 commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -0,0 +1,130 @@
+        jdbcTemplate.update("delete from SECTION where line_id = ? and down_station_id = ?", lineId, stationId);
+        jdbcTemplate.update("insert into SECTION(line_id, up_station_id, down_station_id, distance) VALUES(?,?,?,?)",
+                lineId, downStationSection.getUpStationId(), upStationSection.getDownStationId(), upStationSection.getDistance() + downStationSection.getDistance());
+    }
```
전체적으로 보면 쿼리와 데이터를 통해서 처리하는 부분이 많아보여요.
약간 강제적으로 제안을 드려보자면 select을 제외하고는 update, delete는 id만으로 조작하고  SELECT xxId FROM someTalbe; 과 같은 쿼리는 아이디 조회가 아닌 객체조회로 바꾸고 row를 조회하는 곳에서는 필요한 데이터만 조회하지 말고 모든 컬럼을 조회하여 온전한 객체를 Service 레이어에 전달해주세요.

이미 데이터 중심으로 처리 된 부분이 많아서 개선하고자 하는 service내 코드에서 위 작업을 선행 한 다음 그 이후 객체 중심으로 로직을 작성해보시는 걸 추천드립니다. 

(참고로 리뷰를 코드를 따라가면서 하다보니 피드백 코멘트의 순서와 실제 작성된 순서가 상이할 수 있는데 이게 가장 마지막 작성한 코멘트인 점 참고바래요. 😉)
두번째 질문인 `객체들이 DAO를 직접 가지지 않고서는 이를 구현할 수 있는 방법이 마땅히 떠오르지 않았습니다.`를 어디서 느꼈을 지 알 것 같네요. 객체 단위로 처리를 하기보다는 객체의 상태, 즉 데이터 단위로 집중해서 그런 고민을 하신게 아닌가 싶어요. 아마도 객체의 상태가 바뀌면 그 바뀐 값을 바로 db에 반영하고자 한걸까요? 혹시나 그게 맞다면 객체의 상태를 직접 관리하려고 하지 말고 객체의 상태는 객체 스스로가 관리하고 DB에는 데이터가 아닌 객체의 현재 상태를 반영하는 방향으로 작업해보시길 추천드려요. 🙏

Hue9010 reviewed (changes requested):
안녕하세요. 이번 리뷰를 맡은 휴입니다!
우선 요구사항을 분석하고 설계하는게 어려웠다면 어려운 과정이 맞으니 자연스러운 겁니다. 😉
작업에 따라서는 실제 코드 작성하는 시간보다 분석, 설계에 더 오랜시간이 걸리기도 해요(심지어 분석에만 시간을 가장 많이 쓸수도 있습니다. 😱)

질문 주신 사항은 중간중간 답변드렸는데 변경이 크게 필요해 보이는 부분도 보여서 반영하는데 어렵게 느끼시는 부분 있다면 편히 질문 주세요!

kakao-yell-ow committed:
refactor : 불필요한 메서드 삭제

kakao-yell-ow committed:
refactor : StationService의 stationDao.deleteById 수행 중 오류 발생 시 InvalidStationArgument를 throw하도록 수정

kakao-yell-ow committed:
refactor : StationDao가 InvalidStationArgumentException을 처리하도록 변경

kakao-yell-ow committed:
refactor : StationDao SQL 상수화

kakao-yell-ow committed:
refactor : LineService 리팩터링 및 DAO 주입 방식 변경

DAO 들의 필드 주입을 생성자 주입으로 변경
LineService의 역할 분할을 위해 Sections 일급 컬렉션 생성
LineService의 삽입 / 삭제 시 변경 구간 찾기 기능을 Sections로 이동
메서드 이름 리팩터링 진행

kakao-yell-ow committed:
refactor : 에러 메시지 상수화

kakao-yell-ow committed:
refactor : SQL 쿼리 상수화 및 Dao의 update 인자를 객체로 받도록 수정

kakao-yell-ow committed:
refactor : LineService, LineDao, SectionDao 리팩터링

deleteById의 리턴타입 void로 변경, 삭제 시 삭제된 행이 없는 경우 오류를 리턴하는 부분을 각 Dao가 담당하도록 수정

idealgu commented:
피드백 감사드립니다!
말씀해 주셨던 대로 객체의 상태가 바뀌면 해당 상태를 바로 DB에 저장하려는 의도로 코드를 작성하다보니 
객체 중심적으로 코드를 구현하지 못했던 것 같습니다. 😢
피드백을 참고하여 다음과 같이 리팩터링을 진행해보았습니다.

1. Dao의 쿼리들은 객체를 리턴하도록 수정했습니다.
2. 쿼리에서 조건을 구하며 데이터를 받아오지 않고 List<Section>을 받아와 Sections 일급 컬렉션을 생성하여 그 안에서 구간들의 순서를 구하고, 수정할 구간을 찾는 로직을 구현했습니다.
3. 사용하지 않는 함수들은 확인하여 제거했습니다.

LineService에 있던 기존의 로직이 서로 많이 연관되어 있다보니
리팩터링을 진행하며 Sections 일급 컬렉션을 만들고 서비스의 기능을 위임하려는 과정을 한번에 진행하게 되어
커밋의 단위가 크게 되었습니다..
다른 사람이 보기 어렵지 않을까 걱정이 되는데, 어떤 방식으로 기능 분리를 진행하는 것이 좋은 방식일까요? 😭

Hue9010 commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("lines")
```
👍

Hue9010 commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,86 @@
+        this.lineService = lineService;
+    }
+
+    @ExceptionHandler(InvalidSectionException.class)
```
👍 
flag를 통한 처리를 제거하신 덕분에 비즈니스 코드에선 예외 처리에 대한 코드들이 사라지고 비즈니스 로직에만 집중 하실 수 있게 됐네요. 😉

빠르게 잘 적용해주셨는데 그러면 @ControllerAdvice도 어렵지 않으니 한번 해보시죠.

Hue9010 commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,86 @@
+@RequestMapping("lines")
+public class LineController {
+
+    private LineService lineService;
+
+    public LineController(LineService lineService) {
+        this.lineService = lineService;
+    }
```
생성자를 통한 주입 👍 
이미 보셨을 수도 있지만 생성자 주입을 권장하는 이유입니다. [[Spring]필드 주입(Field Injection) 대신 생성자 주입(Constructor Injection)을 사용해야 하는 이유](https://zorba91.tistory.com/238)

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    private Line saveSectionsTail(Section newSection, Line line) {
+        Line newLine = new Line(line.getId(), line.getStartStationId(), newSection.getDownStationId());
```
Line을 새로 만들 필요가 있을까요? 기존 Line의 downStationId만 변경시켜주면 되지 않을까요? 전체적으로 상태값 변경을 객체를 통해 처리하지 않고 생성자로 새로 만드는데 객체를 통해 자신의 상태를 변경하게 해보세요.

객체지향적인 처리에 대해서는 [생각하라, 객체지향처럼](https://woowabros.github.io/study/2016/07/07/think_object_oriented.html) 보시는 것 추천드려요. 가능하다면 추천 도서에도 있던 객체지향의 사실과 오해를 보는것도 좋답니다. 😉

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+            throw new InvalidSectionException(CONTAINS_BOTH_STATIONS_OR_NOT_EXISTS_STATIONS_ERROR_MESSAGE);
+        }
+        Section newSection = new Section(lineId, sectionRequest.getUpStationId(), sectionRequest.getDownStationId(), sectionRequest.getDistance());
+        if (line.isLineStartStation(sectionRequest.getDownStationId())) {
```
Line 72에서 newSection을 만드셨지만 Line 73에선 다시 sectionRequest의 데이터를 사용하고 계세요.

Service layer 내에서 sectionRequest는 newSection으로 변환 한 다음 사용하지 말아보세요. 😉

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+    public void deleteStationById(Long lineId, Long stationId) {
+        Line line = lineDao.findById(lineId);
+        Sections sections = new Sections(sectionDao.findAllSectionsByLineId(lineId), line.getStartStationId());
+        if (sections.getSize() == 1) {
```
sections의 사이즈가 1인지 확인하기보다는 line.isLineStartStation() 같이 sections한테 삭제가 가능한지 메시지를 보낼 수 있을 것 같아요.

Hue9010 commented on src/main/java/subway/station/StationController.java:
```diff
@@ -1,30 +1,46 @@
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.*;
+import subway.exceptions.DuplicateStationNameException;
+import subway.exceptions.InvalidStationArgumentException;
```
사용되지 않네요.

> MacOS: Ctrl + Opt + o
> Win/Linux: Ctrl + Alt + o
> 자동 설정: [Find Action] -> Optimize imports on 입력 -> “Auto import: …”
[[IntelliJ] intellij 유용한 단축키 정리](https://gmlwjd9405.github.io/2019/05/21/intellij-shortkey.html)

위 단축키나 설정으로 사용하지 않는 import 정리가 가능합니다!

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+        Section deletedSection = sections.findByUpStationId(stationId);
+        Section updatedSection = sections.findByDownStationId(stationId);
+        sectionDao.updateById(new Section(updatedSection.getId(), updatedSection.getLineId(), updatedSection.getUpStationId(),
+                deletedSection.getDownStationId(), updatedSection.getDistance() + deletedSection.getDistance()));
```
이 생성자는 결국 updatedSection의 downStation과 distance를 변경하기 위함인데 생성자를 통하지 말고 메서드를 사용하여 updatedSection의 상태만 변경해보세요.

Hue9010 commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Sections {
```
일급 컬렉션 👍

Hue9010 commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,103 @@
+
+        if (findSection == null) {
+            throw new InvalidSectionException(NO_MATCHING_SECTION_ERROR_MESSAGE);
+        }
```
.orElseThrow()를 활용 할 수 있을 것 같아요!

Hue9010 commented on src/main/java/subway/line/Line.java:
```diff
@@ -0,0 +1,79 @@
+        this.id = id;
+    }
+
+    public boolean isLineStartStation(Long stationId) {
+        return startStationId == stationId;
+    }
```
Line의 startStation인지 확인하는 메서드이기 때문에 사용 될 때 `line.isLineStartStation`이 됩니다. 메서드 명에 line을 넣지 않으셔도 변수명을 a, b로 짓지 않은 이상 line.isStartStation()으로도 충분히 의미 표현이 가능합니다!

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+
+    private Line saveSectionsHead(Section newSection, Line line) {
+        Line newLine = new Line(line.getId(), newSection.getUpStationId(), line.getEndStationId());
+        lineDao.updateLineStartEndStations(newLine);
```
값에 집중하지 않고 객체에 대해서 집중 할 수 있게 한정된 값을 바꾸는 게 아닌 id를 제외한 Line의 모든 상태를 반영하는 dao를 만들어보세요. lineDao.update(line);

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+        return lineDao.findById(newSection.getLineId());
+    }
+
+    private Line saveSectionsTail(Section newSection, Line line) {
```
saveSectionsTail, saveSectionsHead에서
```java
lineDao.updateLineStartStation
lineDao.updateLineEndStation
```

위 두 차이점이 사라지니까 
```java
private void deleteStation(Section section, Line line) {
```
이런 시그니처를 가진 하나의 메서드로 변경 할 여지가 보이네요. 😉

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    public Line updateLine(Long id, LineRequest lineRequest) {
+        return lineDao.updateLineNameAndColor(new Line(id, lineRequest.getName(), lineRequest.getColor()));
```
request -> domain으로 변환하는 과정을 
xxDto.toXX 형태로 변환 메서드를 만들어두면 좀더 간결하게 사용 가능합니다.
ex)
```java
lineRequest.toLine(lineId);
```

Hue9010 reviewed (changes requested):
변경이 많이 필요했을 텐데 피드백 반영 잘해주셨어요. 👏👍
몇가지 더 개선하면 좋을 사항 남겨뒀어요! 궁금하거나 고민되는 사항이 있으면 걱정하지 마시고 좀더 편하게! 문의주시면 돼요. 😏

커밋 히스토리가 다른 사람이 보기 어려울지 너무 고민하지 마세요. 우선 편하신대로 진행해보셔도 충분합니다. 혹 걱정된다면 반영사항을 모두 커밋 메시지에 기술하는 것도 한 방법이에요.😉

kakao-yell-ow committed:
refactor : Line, LineService, Sections 리팩터링

Line의 isLineStartStation, EndStation의 메서드 이름 변경
LineService의 saveSection의 line.isStartStation, EndStation의 파라미터로 newSection의 up/down Station을 넘기게 수정
Sections의 isRemovable 메서드 추가

kakao-yell-ow committed:
refactor : Dto들의 Domain 변환 메서드 추가 및 Sections, LineService 리팩터링

Sections의 isRemovable의 부등호 삭제
LineRequest, SectionRequest의 toLine, toSection 메서드 추가
Dto 변환 메서드 추가에 따른 LineService 리팩터링

kakao-yell-ow committed:
refactor : 객체 상태 업데이트 리팩터링 및 null처리 방식 리팩터링

Sections의 StationId를 찾는 Stream에서 찾지 못한 경우 orElseThrow로 Exception을 throw하도록 수정
Section에 삭제/수정에 따른 기존 Section 객체의 상태를 변경하는 updateSectionInfoWhenInserted/Deleted 메서드 추가
Line에 삭제/수정에 따른 기존 Line의 Start/End Station 상태를 변경하는 updateStationInfoWhenInserted/Deleted 메서드 추가
Line에 LineRequest에 따라 이름과 색깔을 변경하는 updateLineInfo 메서드 추가
Line, Section, Sections 리팩터링에 따른 LineService 로직 리팩터링

kakao-yell-ow committed:
refactor : 사용하지 않는 import 삭제 및 ControllerAdvice로 Error Handling 하도록 수정

kakao-yell-ow committed:
refactor : 사용하지 않는 메서드 제거

kakao-yell-ow committed:
refactor : LineDao의 updateLine 메서드 명을 updateById로 변경

idealgu commented:
많은 피드백 감사드립니다!! 😁
코드를 작성하면서도 new로 매번 업데이트할 객체를 새로 만들어주는 방식에 대해 이렇게 구현해도 되나 하는 의문이 있었는데
말씀해주신대로 객체에 메시지를 보내며 업데이트 하는 방식으로 교체하는 방식이 더욱 객체지향적으로 처리가 가능하게 된 것 같습니다!
그 밖에 말씀해 주셨던 부분들에 대해서도 리팩토링을 진행해보았습니다. 😃

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,129 @@
+
+    public Line updateLine(Long id, LineRequest lineRequest) {
+        Line line = lineDao.findById(id);
+        line.updateLineInfo(lineRequest);
```
비효율 적으로 느낄 수 있겠지만 도메인 간의 협력을 통해 상태 변화가 이뤄지게 DTO가 아닌 Line을 전달해주세요. 혹은 특정 값을 line한테 전달 해 줄 수도 있을 것 같아요.
추가로 updateLineInfo를 통해 Line이 LineRequest에 의존하는 부분이 생겼네요.

isangu replied:
사실 처음에 구현은 updateLineInfo(String name, String color)로 이름과 색깔 정보를 변경하려고 했는데,
setter 함수를 써서 변경하는 것으로 느껴져서 lineRequest를 받도록 수정했었습니다..😅
원래 구현처럼 updateLineInfo(String name, String color)와 같이 구현하는 것이 더 좋은 방식일까요?

Hue9010 replied:
넵 혹은 lineRequest를 Line으로 변환 한다음에 line.update(newLine)으로도 갱신가능할 것 같아요.

Hue9010 commented on src/main/java/subway/line/Line.java:
```diff
@@ -0,0 +1,104 @@
+        this.color = lineRequest.getColor();
+    }
+
+    public void updateStationInfoWhenInserted(Section section) {
+        if(section.getDownStationId() == startStationId) {
+            this.startStationId = section.getUpStationId();
+        }
+        if(section.getUpStationId() == endStationId) {
+            this.endStationId = section.getDownStationId();
+        }
+    }
```
Line에 속한 Section이 추가되는거니까 addSection은 어떨까요? 🤔 
```java
line.addSection(section);
```
updateStationInfoWhenDeleted도 마찬가지!

isangu replied:
말씀해주신것처럼 addSection/deleteSection이 더 적절한 메서드 명인것 같아 교체했습니다! 😄

Hue9010 commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,99 @@
+        return sections.get(sections.size() - 1).getDownStationId();
+    }
+
+    public int getSize() {
+        return sections.size();
+    }
```
사용하지 않는 메서드는 그때 그때 지워주세요! 😉

Hue9010 commented on src/main/java/subway/advice/ExceptionAdvice.java:
```diff
@@ -0,0 +1,27 @@
+import subway.exceptions.*;
+
+@ControllerAdvice
+public class ExceptionAdvice {
```
잘해주셨네요. 👍  생각보다 쉽죠? 😉 
ControllerAdvice 덕분에 비즈니스 로직을 작성하는 곳에서는 에러 처리가 아닌 로직에 집중할 수 있게 됩니다.
이참에 AOP라는 키워드를 학습한다기보다는 가볍게 봐두시면 좋을 것 같아요. 😉

Hue9010 reviewed (approved):
잘 반영해주셨네요! 💯
이번 단계는 여기서 머지하도록 할게요. 다음 단계 진행하시면 됩니다!
궁금한 부분은 pr 이전이라도 언제든 질문 주세요. 😉
